src/main/java/gregtech/api/recipes/RecipeMap.java

```diff
@@ -228,46 +249,106 @@ public Recipe findRecipe(long voltage, IItemHandlerModifiable inputs, IMultipleT
     @Nullable
     public Recipe findRecipe(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, int outputFluidTankCapacity, MatchingMode matchingMode) {
-        if (recipeList.isEmpty())
+        return findRecipe(voltage, inputs, fluidInputs, outputFluidTankCapacity, matchingMode, false);
+    }
+
+    /**
+     * Finds a Recipe matching the Fluid and/or ItemStack Inputs.
+     *
+     * @param voltage                 Voltage of the Machine or Long.MAX_VALUE if it has no Voltage
+     * @param inputs                  the Item Inputs
+     * @param fluidInputs             the Fluid Inputs
+     * @param outputFluidTankCapacity minimal capacity of output fluid tank, used for fluid canner recipes for example
+     * @param matchingMode            matching logic used for finding the recipe according to {@link MatchingMode}
+     * @param exactVoltage            should require exact voltage matching on recipe. used by craftweaker
+     * @return the Recipe it has found or null for no matching Recipe
+     */
+
+    @Nullable
+    public Recipe findRecipe(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, int outputFluidTankCapacity, MatchingMode matchingMode, boolean exactVoltage) {
+        if (recipeSet.isEmpty())
             return null;
         if (minFluidInputs > 0 && GTUtility.amountOfNonNullElements(fluidInputs) < minFluidInputs) {
             return null;
         }
         if (minInputs > 0 && GTUtility.amountOfNonEmptyStacks(inputs) < minInputs) {
             return null;
         }
-        if (maxInputs > 0) {
-            return findByInputs(voltage, inputs, fluidInputs, matchingMode);
-        } else {
-            return findByFluidInputs(voltage, inputs, fluidInputs, matchingMode);
-        }
+        return findByInputsAndFluids(voltage, inputs, fluidInputs, matchingMode, exactVoltage);
     }
 
     @Nullable
-    private Recipe findByFluidInputs(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, MatchingMode matchingMode) {
-        for (FluidStack fluid : fluidInputs) {
-            if (fluid == null) continue;
-            Collection<Recipe> recipes = recipeFluidMap.get(new FluidKey(fluid));
-            if (recipes == null) continue;
-            for (Recipe tmpRecipe : recipes) {
-                if (tmpRecipe.matches(false, inputs, fluidInputs, matchingMode)) {
-                    return voltage >= tmpRecipe.getEUt() ? tmpRecipe : null;
+    private Recipe findByInputsAndFluids(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, MatchingMode matchingMode, boolean exactVoltage) {
+        HashSet<Recipe> iteratedRecipes = new HashSet<>();
+        HashSet<ItemStackKey> searchedItems = new HashSet<>();
+        HashSet<FluidKey> searchedFluids = new HashSet<>();
+        HashMap<Integer, LinkedList<Recipe>> priorityRecipeMap = new HashMap<>();
+        HashMap<Recipe, Integer> promotedTimes = new HashMap<>();
+
+        for (ItemStack stack : inputs) {
```

Comment:
Can we short circuit this recipe promotion even more by checking if `matchingMode` is set to `IGNORE_ITEMS` and skipping the promotion if it is?